src/main/kotlin/org/koil/user/register/RegistrationController.kt

```diff
@@ -41,23 +69,72 @@ data class RegistrationAttempt(
 @RequestMapping("/auth")
 class RegistrationController(
     @Autowired private val users: UserService,
+    @Autowired private val companies: OrganizationService
 ) {
-    @GetMapping("/register")
-    fun register(
+    @GetMapping("/register/individual")
+    fun registerIndividual(
         @AuthenticationPrincipal user: EnrichedUserDetails?,
-        @RequestParam("email", defaultValue = "") email: String
+        @RequestParam("email", defaultValue = "") email: String,
+        @RequestParam("signupLink") signupLink: UUID?,
     ): ModelAndView {
-        return if (user == null) {
-            RegisterViews.Register.render(RegistrationViewModel(email))
+        return if (user != null) {
+            ModelAndView("redirect:/dashboard")
+        } else if (signupLink == null) {
+            ModelAndView("redirect:/auth/register/organization?email=$email")
         } else {
+            RegisterViews.RegisterIndividual.render(RegistrationViewModel(email, signupLink = signupLink))
+        }
+    }
+
+    @GetMapping("/register/organization")
+    fun registerOrganization(
+        @AuthenticationPrincipal user: EnrichedUserDetails?,
+        @RequestParam("email", defaultValue = "") email: String,
+    ): ModelAndView {
+        return if (user != null) {
             ModelAndView("redirect:/dashboard")
+        } else {
+            RegisterViews.RegsterOrganization.render(OrganizationRegistrationViewModel(email))
+        }
+    }
+
+    @PostMapping("/register/organization")
+    fun registerOrganizationSubmit(
+        request: HttpServletRequest,
+        @Valid @ModelAttribute("submitted") submitted: OrganizationRegistrationAttempt,
+        result: BindingResult
+    ): ModelAndView {
+        return if (!result.hasErrors()) {
+            when (companies.setupOrganization(submitted.toSetupRequest())) {
+                is OrganizationCreatedResult.CreatedOrganization -> {
+                    request.login(submitted.email, submitted.password)
+                    ModelAndView("redirect:/dashboard")
+                }
+                is OrganizationCreatedResult.UserCreationFailed -> {
+                    RegisterViews.RegsterOrganization.render(
+                        OrganizationRegistrationViewModel(email = submitted.email, emailAlreadyTaken = true),
+                        HttpStatus.BAD_REQUEST
+                    )
+                }
+                is OrganizationCreatedResult.CreationFailed -> {
```

Comment:
I wonder if this generic failure can just be an exception. We don't really expect this to happen.